encoders/firebase-decoders-json/src/test/java/com/google/firebase/decoders/TypeTokenTest.java

```diff
@@ -0,0 +1,167 @@
+    TypeToken.ClassToken<Map<String, Foo>> mapClassToken =
+        (TypeToken.ClassToken<Map<String, Foo>>) mapTypeToken;
+    TypeTokenContainer typeTokenContainer = mapClassToken.getTypeArguments();
+    TypeToken<String> firstArgumentToken = typeTokenContainer.at(0);
+    TypeToken<Foo> secondArgumentTypeToken = typeTokenContainer.at(1);
+    assertThat(firstArgumentToken).isInstanceOf(TypeToken.ClassToken.class);
+    assertThat(secondArgumentTypeToken).isInstanceOf(TypeToken.ClassToken.class);
+    TypeToken.ClassToken<String> firstArgumentClassToken =
+        (TypeToken.ClassToken<String>) firstArgumentToken;
+    TypeToken.ClassToken<Foo> secondArgumentClassToken =
+        (TypeToken.ClassToken<Foo>) secondArgumentTypeToken;
+    assertThat(firstArgumentClassToken.getRawType()).isEqualTo(String.class);
+    assertThat(secondArgumentClassToken.getRawType()).isEqualTo(Foo.class);
+  }
+
+  @Test
+  public void nestedGenericType_actualTypeParametersAreCorrectlyCaptured() {
+    TypeToken<List<List<String>>> typeToken = TypeToken.of(new Safe<List<List<String>>>() {});
+    assertThat(typeToken).isInstanceOf(TypeToken.ClassToken.class);
+    TypeToken.ClassToken<List<List<String>>> typeClassToken =
+        (TypeToken.ClassToken<List<List<String>>>) typeToken;
+    TypeToken<List<String>> componentToken = typeClassToken.getTypeArguments().at(0);
+    assertThat(componentToken).isInstanceOf(TypeToken.ClassToken.class);
+    TypeToken.ClassToken<List<String>> componentClassToken =
+        (TypeToken.ClassToken<List<String>>) componentToken;
+    assertThat(componentClassToken.getRawType()).isEqualTo(List.class);
+    TypeToken<String> innerComponentToken = componentClassToken.getTypeArguments().at(0);
+    assertThat(innerComponentToken).isInstanceOf(TypeToken.ClassToken.class);
+    TypeToken.ClassToken<String> innerComponentClassToken =
+        (TypeToken.ClassToken<String>) innerComponentToken;
+    assertThat(innerComponentClassToken.getRawType()).isEqualTo(String.class);
+  }
+
+  // Bounded Wildcard Type
+  @Test
+  public void boundedWildcardTypeWithExtend_actualTypeParameterIsCastedToUpperBound() {
+    TypeToken<List<? extends Number>> typeToken =
+        TypeToken.of(new Safe<List<? extends Number>>() {});
+    assertThat(typeToken).isInstanceOf(TypeToken.ClassToken.class);
+    TypeToken.ClassToken<List<? extends Number>> typeClassToken =
+        (TypeToken.ClassToken<List<? extends Number>>) typeToken;
+    TypeToken<Number> componentType = typeClassToken.getTypeArguments().at(0);
+    assertThat(componentType).isInstanceOf(TypeToken.ClassToken.class);
+    TypeToken.ClassToken<Number> componentClassType = (TypeToken.ClassToken<Number>) componentType;
+    assertThat(componentClassType.getRawType()).isEqualTo(Number.class);
+  }
+
+  @Test
+  public void boundedWildcardTypeWithSuper_notSupported() {
+    TypeToken<List<? super Foo>> typeToken = TypeToken.of(new Safe<List<? super Foo>>() {});
```

Comment:
In the spirit of https://en.wikipedia.org/wiki/Fail-fast, Ideally the exception should be thrown at this point, not when one tries to get the type arguments of a type token. Feel free to address this in a separate PR.